File: src/main/java/com/github/njuro/jard/ban/BanFacade.java
package com.github.njuro.jard.ban;

import com.github.njuro.jard.ban.dto.BanDto;
import com.github.njuro.jard.ban.dto.BanForm;
import com.github.njuro.jard.base.BaseFacade;
import com.github.njuro.jard.common.Constants;
import com.github.njuro.jard.user.UserFacade;
import com.github.njuro.jard.user.dto.UserDto;
import com.github.njuro.jard.utils.validation.PropertyValidationException;
import java.time.OffsetDateTime;
import java.util.Comparator;
import java.util.List;
import java.util.UUID;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.scheduling.annotation.Scheduled;
import org.springframework.stereotype.Component;

@Component
public class BanFacade extends BaseFacade<Ban, BanDto> {

  private final UserFacade userFacade;

  private final BanService banService;

  @Autowired
  public BanFacade(UserFacade userFacade, BanService banService) {
    this.userFacade = userFacade;
    this.banService = banService;
  }

  /**
   * Creates and saves new ban or warning. New ban is active immediately.
   *
   * @param banForm form with ban data
   * @return created ban
   * @throws PropertyValidationException if no user is logged in or there is already active ban on
   *     given IP
   */
  public BanDto createBan(BanForm banForm) {
    UserDto loggedUser = userFacade.getCurrentUser();
    if (loggedUser == null) {
      throw new PropertyValidationException("No user is logged in");
    }

    if (banService.hasActiveBan(banForm.getIp())) {
      throw new PropertyValidationException("There is already active ban on this IP");
    }

    BanDto ban = banForm.toDto();
    ban.setBannedBy(loggedUser);
    ban.setValidFrom(OffsetDateTime.now());

    if (ban.getStatus() == BanStatus.WARNING) {
      ban.setValidTo(null);
    }

    return toDto(banService.saveBan(toEntity(ban)));
  }

  /** {@link BanService#getActiveBan(String)} */
  public BanDto getActiveBan(String ip) {
    return toDto(banService.getActiveBan(ip));
  }

  /** {@link BanService#hasActiveBan(String)} */
  public boolean hasActiveBan(String ip) {
    return banService.hasActiveBan(ip);
  }

  /** @return all bans sorted by most to least recent */
  public List<BanDto> getAllBans() {
    List<Ban> bans = banService.getAllBans();
    bans.sort(Comparator.comparing(Ban::getValidFrom).reversed());
    return toDtoList(bans);
  }

  /** {@link BanService#resolveBan(UUID)} */
  public BanDto resolveBan(UUID id) {
    return toDto(banService.resolveBan(id));
  }

  /**
   * Edits a ban. Only reason and ban ending data can be edited.
   *
   * @param oldBan ban to be edited
   * @param banForm form with new values
   * @return updated ban
   */
  public BanDto editBan(BanDto oldBan, BanForm banForm) {
    oldBan.setReason(banForm.getReason());
    oldBan.setValidTo(banForm.getValidTo());

    return toDto(banService.saveBan(toEntity(oldBan)));
  }

  /**
   * Invalidates active ban.
   *
   * @param ban ban to invalidate
   * @param unbanForm form with details of invalidation
   * @return invalidated ban
   * @throws PropertyValidationException if no user is logged in or there is no active ban on given
   *     IP
   */
  public BanDto unban(BanDto ban, UnbanForm unbanForm) {
    UserDto loggedUser = userFacade.getCurrentUser();
    if (loggedUser == null) {
      throw new PropertyValidationException("No user is logged in!");
    }

    if (ban == null || ban.getStatus() != BanStatus.ACTIVE) {
      throw new PropertyValidationException("There is no active ban on this IP");
    }

    ban.setUnbannedBy(loggedUser);
    ban.setUnbanReason(unbanForm.getReason());
    ban.setStatus(BanStatus.UNBANNED);

    return toDto(banService.saveBan(toEntity(ban)));
  }

  /**
   * Periodically checks for expired active bans and invalidates them. Period is defined by {@link
   * Constants#EXPIRED_BANS_CHECK_PERIOD}.
   */
  @Scheduled(fixedRateString = Constants.EXPIRED_BANS_CHECK_PERIOD)
  public void unbanExpired() {
    List<Ban> expiredBans = banService.getExpiredBans();
    expiredBans.forEach(
        ban -> {
          ban.setStatus(BanStatus.EXPIRED);
          banService.saveBan(ban);
        });
  }
}

File: src/main/kotlin/com/github/njuro/jard/rewrite/ban/BanFacade.kt
package com.github.njuro.jard.rewrite.ban

import com.github.njuro.jard.rewrite.ban.BanStatus.EXPIRED
import com.github.njuro.jard.rewrite.ban.BanStatus.UNBANNED
import com.github.njuro.jard.rewrite.ban.BanStatus.WARNING
import com.github.njuro.jard.rewrite.ban.dto.BanDto
import com.github.njuro.jard.rewrite.ban.dto.BanForm
import com.github.njuro.jard.rewrite.ban.dto.UnbanForm
import com.github.njuro.jard.rewrite.base.BaseFacade
import com.github.njuro.jard.rewrite.common.EXPIRED_BANS_CHECK_PERIOD
import com.github.njuro.jard.rewrite.user.UserFacade
import com.github.njuro.jard.rewrite.utils.validation.PropertyValidationException
import org.springframework.scheduling.annotation.Scheduled
import org.springframework.stereotype.Component
import java.time.OffsetDateTime
import java.util.UUID

@Component
class BanFacade(
    private val userFacade: UserFacade,
    private val banService: BanService
    ) : BaseFacade<Ban, BanDto>() {
    /**
     * Creates and saves new ban or warning. New ban is active immediately.
     *
     * @param banForm form with ban data
     * @return created ban
     * @throws PropertyValidationException if no user is logged in or there is already active ban on
     * given IP
     */
    fun createBan(banForm: BanForm): BanDto {
        val loggedUser = userFacade.getCurrentUser() ?: throw PropertyValidationException("No user is logged in")
        if (banService.hasActiveBan(banForm.ip)) {
            throw PropertyValidationException("There is already active ban on this IP")
        }
        val ban = banForm.toDto().let {
            it.copy(
                bannedBy = loggedUser,
                validFrom = OffsetDateTime.now(),
                validTo = if (it.status == WARNING) null else it.validTo
            )
        }
        return toDto(banService.saveBan(toEntity(ban)))
    }

    /** [BanService.getActiveBan]  */
    fun getActiveBan(ip: String): BanDto =
        toDto(banService.getActiveBan(ip) ?: throw BanNotFoundException())


    /** [BanService.hasActiveBan]  */
    fun hasActiveBan(ip: String): Boolean = banService.hasActiveBan(ip)


    /** @return all bans sorted by most to least recent
     */
    fun getAllBans(): List<BanDto> {
        val bans = banService.getAllBans().sortedByDescending(Ban::validFrom)
        return toDtoList(bans)
    }

    /** [BanService.resolveBan]  */
    fun resolveBan(id: UUID): BanDto? = toDto(banService.resolveBan(id))


    /**
     * Edits a ban. Only reason and ban ending data can be edited.
     *
     * @param oldBan ban to be edited
     * @param banForm form with new values
     * @return updated ban
     */
    fun editBan(oldBan: BanDto, banForm: BanForm): BanDto {
        val updated = oldBan.copy(reason = banForm.reason, validTo = banForm.validTo)
        return toDto(banService.saveBan(toEntity(updated)))
    }

    /**
     * Invalidates active ban.
     *
     * @param ban ban to invalidate
     * @param unbanForm form with details of invalidation
     * @return invalidated ban
     * @throws PropertyValidationException if no user is logged in or there is no active ban on given
     * IP
     */
    fun unban(ban: BanDto, unbanForm: UnbanForm): BanDto {
        val loggedUser = userFacade.getCurrentUser() ?: throw PropertyValidationException("No user is logged in!")
        if (ban.status !== BanStatus.ACTIVE) {
            throw PropertyValidationException("There is no active ban on this IP")
        }
        val updated = ban.copy(
            unbannedBy = loggedUser,
            unbanReason = unbanForm.reason,
            status = UNBANNED
        )
        return toDto(banService.saveBan(toEntity(updated)))
    }

    /**
     * Periodically checks for expired active bans and invalidates them. Period is defined by [EXPIRED_BANS_CHECK_PERIOD].
     */
    @Scheduled(fixedRateString = EXPIRED_BANS_CHECK_PERIOD)
    fun unbanExpired() {
        banService.getExpiredBans().forEach {
            it.status = EXPIRED
            banService.saveBan(it)
        }
    }
}
